doc/man/man3/notcurses.3.md
% notcurses(3)
% nick black <nickblack@linux.com>
% v3.0.17

# NAME

notcurses - TUI library for modern terminal emulators

# SYNOPSIS

**#include <notcurses/notcurses.h>** or
**#include <notcurses/notcurses-core.h>**

**-lnotcurses-core -lnotcurses** or **-lnotcurses-core**

# DESCRIPTION

Notcurses builds atop the **terminfo(5)** abstraction layer to provide
reasonably portable vivid character displays. It is an intellectual descendant
of **ncurses(3NCURSES)**, but goes beyond that library (and the X/Open Curses
API it implements).

A program wishing to use Notcurses will need to link it, ideally using the
output of **pkg-config --libs notcurses** (see **pkg-config(1)**). It is
advised to compile with the output of **pkg-config --cflags notcurses**. If
using CMake, a support file is provided, and can be accessed as **Notcurses**
(see **cmake(1)**). If multimedia capabilities are not needed, it is possible
to link against a minimal Notcurses using **pkg-config --libs notcurses-core**.

**notcurses_init(3)** can then be used to initialize a Notcurses instance for a
given **FILE*** (usually **stdout**, usually attached to a terminal).

## The alternate screen

Many terminals provide an "alternate screen" with its own contents, no
scrollback buffer, and no scrolling. Entering the alternate screen replaces
the current visible contents wholesale, as does returning to the regular
screen. Notcurses refers to the alternate screen's semantics as "TUI mode",
and the regular screen's semantics as "CLI mode". It is possible to swap
between the two modes at runtime using **notcurses_leave_alternate_screen(3)**
and **notcurses_enter_alternate_screen(3)**. Notcurses will enter TUI mode
by default on startup; to prevent this, use **NCOPTION_NO_ALTERNATE_SCREEN**
as described in **notcurses_init(3)**. On program exit, Notcurses will always
return to the regular screen, independent of the screen being used on
program start.

## Construction

Before calling into Notcurses—and usually as one of the first calls of the
program—be sure to call **setlocale** with an appropriate UTF-8 **LC_ALL**
locale. It is usually appropriate to use **setlocale(LC_ALL, "")**, relying on
the user to properly set the **LANG** environment variable. Notcurses will
refuse to start if **nl_langinfo(3)** doesn't indicate UTF-8 or ANSI_X3.4-1968
(aka US-ASCII). Be aware that capabilities are substantially reduced in ASCII.

**notcurses_init(3)** accepts a **struct notcurses_options** allowing fine-grained
control of Notcurses behavior, including signal handlers, alternative screens,
and overriding the TERM environment variable. A **terminfo** entry appropriate
for the actual terminal must be available.

**ncdirect_init(3)** makes available a restricted subset of Notcurses
functionality. This subset is intended to be interleaved with user-generated
output, and is limited to coloring and styling. Direct mode is documented in
**notcurses_direct(3)**.

Only one context can be active in a process at a time, whether direct mode
(**struct ncdirect**) or rendered mode (**struct notcurses**).

## Output

All output is performed on **struct ncplane**s (see [Ncplanes][] below). Output
is not visible until explicitly rendered via **notcurses_render(3)**.
Information on drawing functions is available at **notcurses_output(3)**.

## Input

Notcurses supports input from keyboards (via **stdin**) and pointing devices (via
a broker such as GPM, X, or Wayland). Input is delivered as 32-bit Unicode
code points. Synthesized events such as mouse button presses and arrow keys
are mapped into Unicode's
[Supplementary Private Use Area-B](https://unicode.org/charts/PDF/U100000.pdf).
Information on input is available at **notcurses_input(3)**. The included tool
**notcurses-input(1)** can be used to test input decoding.

## Ncpiles

A given **notcurses** context is made up of one or more piles. Piles provide
distinct rendering contexts: a thread can be rendering or mutating one pile,
while another thread concurrently renders or mutates another pile. A pile is
made up of planes, totally ordered on a z-axis. In addition to the z-ordering,
the planes of a pile are bound in a forest (a set of directed, acyclic graphs).
Those planes which are not bound to some other plane constitute the root planes
of a pile. A pile is destroyed when all its planes are destroyed, or moved to
other piles. Since the standard plane (see below) always exists, and cannot be
moved to another pile, one pile always exists, known as the standard pile.

Note that rasterizing a pile will replace all content within its margins.

For more information, see **notcurses_pile(3)**.

## Ncplanes

Following initialization, a single ncplane exists, the "standard plane" (see
**notcurses_stdplane(3)**). This plane cannot be destroyed nor manually resized,
and is always exactly as large as the screen (if run without a TTY, the "screen"
is assumed to be 80x24 cells). Further ncplanes can be created with
**ncplane_create(3)**. A total z-ordering always exists on the set of ncplanes,
and new ncplanes are placed at the top of the z-buffer. Ncplanes can be larger,
smaller, or the same size as the physical screen, and can be placed anywhere
relative to it (including entirely off-screen). Ncplanes are made up of
**nccell**s (see [NcCells][] below). Information on ncplanes is available at
**notcurses_plane(3)**.

## NcCells

**nccell**s make up the framebuffers backing each ncplane, one cell per
coordinate, one extended grapheme cluster (see **unicode(7)**) per cell. An
**nccell** consists of a gcluster (either a directly-encoded 7-bit ASCII
character (see **ascii(7)**), or a 25-bit index into the ncplane's egcpool), a
set of attributes, and two channels (one for the foreground, and one for the
background—see **notcurses_channels(3)**). Information on cells is available at
**notcurses_cell(3)**.

It is not usually necessary for users to interact directly with **nccell**s. They
are typically encountered when retrieving data from ncplanes or the rendered
scene (see e.g. **ncplane_at_yx(3)**), or to achieve peak performance when a
particular EGC is heavily reused within a plane.

## Visuals

Bitmaps can be loaded from disk or memory, or even synthesized from the
content of existing planes. These are stored in **ncvisual** objects, described
in **notcurses_visual(3)**. Visuals can be rendered to arbitrarily many
planes using a variety of blitters, varying in their aspect ratios and
resolution. If the terminal supports a pixel protocol such as Sixel or
Kitty, it is possible to render bitmaps at the pixel level (as opposed to
the cell level, using geometric glyphs). Otherwise, various Unicode-based
blitters are available to render bitmaps in the text paradigm.

## Widgets

A few high-level widgets are included, all built atop ncplanes:

* **notcurses_fds(3)** for dumping file descriptors/subprocesses to a plane
* **notcurses_menu(3)** for menu bars at the top or bottom of the screen
* **notcurses_multiselector(3)** for selecting one or more items from a set
* **notcurses_plot(3)** for drawing histograms and lineplots
* **notcurses_progbar(3)** for drawing progress bars
* **notcurses_reader(3)** for free-form input data
* **notcurses_reel(3)** for hierarchal display of block-based data
* **notcurses_tabbed(3)** for tabbed interfaces
* **notcurses_selector(3)** for selecting one item from a set
* **notcurses_tree(3)** for hierarchal display of line-based data

## Threads

Notcurses explicitly supports use in multithreaded environments, but it does
not itself perform any locking.

* Only one pile's rendered frame can be rasterized at a time, and it is **not**
  safe to concurrently render that pile. It is safe to rasterize a frame while
  rendering some other pile.
* It is otherwise always safe to operate concurrently on distinct piles.
* It is not safe to render a pile while concurrently modifying that pile.
* It is safe to output to multiple distinct ncplanes at the same time, even
  within the same pile.
* It is safe to output to ncplanes while adding or deleting some other ncplane.
* It is **not** safe for multiple threads to output to the same ncplane.
* It is **not** safe to add, delete, or reorder ncplanes within a single pile
  from multiple threads.

Only one thread may call **notcurses_get(3)** or any other input-related thread
at a time, but it **is** safe to call for input while another thread renders.

Since multiple threads can concurrently manipulate distinct ncplanes, peak
performance might require dividing the screen into several planes, and
manipulating them from multiple threads.

## Destruction

Before exiting, **notcurses_stop(3)** should be called. In addition to freeing up
resources, this is necessary to restore the terminal to a state useful for the
shell. By default, **notcurses_init(3)** installs signal handlers to catch all
signals which would normally terminate the process. The new handlers will try
to call **notcurses_stop(3)**, and then propagate the received signal to the
previous action.

# ENVIRONMENT VARIABLES

The **TERM** environment variable ought be correctly defined. It will be used
to index into the **terminfo(5)** database by way of **setupterm(3NCURSES)**.
Notcurses will additionally use **TERM_PROGRAM** to distinguish certain
terminals.

If the **COLORTERM** environment variable is defined as "**24bit**" or
"**truecolor**", Notcurses will assume the terminal capable of 24-bit RGB
color, even in the absence of "**RGB**" or "**Tc**" capabilities in
terminfo.

If the **NOTCURSES_LOGLEVEL** environment variable is defined as a number
between -1 and 8, inclusive, that will override any logging level specified
in the **struct notcurses_options** provided to **notcurses_init(3)**.

The **LOGNAME** environment variable, if defined, will be used for
**notcurses_accountname(3)**.

# NOTES

When using the C++ wrappers, **NCPP_EXCEPTIONS_PLEASE** can be defined in
order to turn most error returns into exceptions.

# SEE ALSO

**ncurses(3NCURSES)**,
**notcurses-demo(1)**,
**notcurses-input(1)**,
**notcurses_capabilities(3)**,
**notcurses_cell(3)**,
**notcurses_channels(3)**,
**notcurses_direct(3)**,
**notcurses_fade(3)**,
**notcurses_fds(3)**,
**notcurses_init(3)**,
**notcurses_input(3)**,
**notcurses_lines(3)**,
**notcurses_menu(3)**,
**notcurses_multiselector(3)**,
**notcurses_output(3)**,
**notcurses_palette(3)**,
**notcurses_pile(3)**,
**notcurses_plane(3)**,
**notcurses_plot(3)**,
**notcurses_progbar(3)**,
**notcurses_reader(3)**,
**notcurses_reel(3)**,
**notcurses_refresh(3)**,
**notcurses_render(3)**,
**notcurses_selector(3)**,
**notcurses_stats(3)**,
**notcurses_stdplane(3)**,
**notcurses_stop(3)**,
**notcurses_tabbed(3)**,
**notcurses_tree(3)**,
**notcurses_visual(3)**,
**terminfo(5)**, **ascii(7)**, **utf-8(7)**,
**unicode(7)**


NEWS.md
This document attempts to list user-visible changes and any major internal
rearrangements of Notcurses.

* 3.0.17 (2025-10-28)
  * Fix build problems on Windows and Mac OSX.

* 3.0.16 (2025-05-04)
  * Fix several bugs in input handling that could lead to invalid
    memory reads (#2875)

* 3.0.15 (2025-04-29)
  * Fix bug on FreeBSD where we could stop processing input (#2873).

* 3.0.14 (2025-04-26)
  * `ncplane_family_destroy()` has been added to the API.
  * Added some `foot` capabilities. Recognize `ghostty` and bless its
     quadrants/sextants implementations.
  * A bug introduced sometime in 2022 that caused unpredictable
    hangs on exit was resolved (#2837), yay!
  * Fix bug identified by @kmarius on github where we core out if printing
     a tab at the end of an ncplane whose x dimension is not a multiple
     of the tab stop (#2870), thanks for the report!
  * We now require CMake 3.21 (vs 3.14) and C17 (vs C11).

* 3.0.13 (2025-01-11)
  * Fix regression when building with `USE_CXX=off`.
  * Use `distutils` from its own Python component rather than assuming
    it's in `setuputils`, from which it was removed in Python 3.12.
  * Properly check for UTF8 before calling ncmetric_use_utf8(), fixing
    a bug when using ncmetric without UTF8 support.

* 3.0.12 (2025-01-09)
  * Fixed a bug when rendering QR codes into a small area. QR codes now
    require `NCBLIT_2x1`, as that is the only blitter which can generate a
    proper aspect ratio. It thus no longer works in an ASCII environment.
  * The `NCBLIT_4x2` "octant" blitter has been added, making use of new
    characters from Unicode 16. `notcurses_canoctant()` has been added to
    check for `NCBLIT_4x2` support at runtime. If present, octants will be
    used for `NCBLIT_DEFAULT` when used with `NCSCALE_NONE_HIRES`,
    `NCSCALE_SCALE_HIRES`, or `NCSCALE_STRETCH`. Thanks, eschnett! Note
    that octants are not supported by GNU libc until 2.41.
  * Fixed coredump on exit when using musl as libc (alpine, some gentoo).

* 3.0.11 (2024-10-02)
  * We now normalize the return of `nl_langinfo()` according to the behavior
    of glibc's `_nl_normalize_charset()`, supporting some atypical synonyms
    of `UTF-8`.
  * Fixed a bug in `ncsixel_as_rgba()` (called by `ncvisual_from_sixel()`)
    that broke loading of sixels having more than 12 rows (sixel generation
    from images worked fine). Thanks, waveplate!
  * Reject illegal geometries in `ncvisual_from_*()`.
  * We build with FFMPEG 7.1.

* 3.0.10 (2024-10-02)
  * Cursed. See https://github.com/dankamongmen/notcurses/issues/2795.

* 3.0.9 (2022-12-10)
  * Eliminated infinite loop in `ncplane_move_family_above()`.
    Thanks, drewt!

* 3.0.8 (2022-04-06)
  * Bugfix release, but also support curled underlines in Contour.

* 3.0.7 (2022-02-20)
  * Tab characters may now be used with `ncplane_put*()`. See
     `ncplane_output.3` for more information.

* 3.0.6 (2022-02-09)
  * `ncplane_set_[fb]channel()`, `ncplane_[fb]channel()`,
    `ncplane_channels()`, `ncplane_set_channels()`,
    `ncchannels_set_[fb]channel()`, and `ncchannels_[fb]channel()` now
    function only on the 28 alpha + coloring bits of their respective
    channels, which is almost certainly what you wanted in the first place.
  * Restore `nccell_set_[fb]channel()` and friends, using these semantics.

* 3.0.5 (2022-01-21)
  * The Hyper and Super modifiers are now supported. CapsLock and NumLock
    can further be disambiguated when using the Kitty protocol. The
    new functions `ncinput_super_p()` and friends have been added.
  * `ncinput` has a new field, `modifiers`. The old `alt`, `shift`, and
    `ctrl` booleans are now deprecated, and will be removed in 4.0.
  * `ncmenu_section` **must** now specify any expected modifiers for their
    shortucts using `modifiers`. Setting any of `alt`, `shift`, or
    `ctrl` will see `ncmenu_create()` fail.
  * `ncinput_equal_p()` considers `NCTYPE_UNKNOWN` equal to `NCTYPE_PRESS`.
  * Added `ncpalette_get()` for orthogonality's sake.

* 3.0.4 (2022-01-08)
  * We now use level 2 of `XTMODKEYS`, providing better differentiation
    of keyboard modifiers. We now unpack the Meta modifier.
  * Added `ncinput_shift_p()`, `ncinput_alt_p()`, `ncinput_ctrl_p()`,
    and `ncinput_meta_p()` to test for various modifiers in `ncinput`s.
  * When support is advertised, the 1016 mouse protocol will be used
    to provide pixel-level detail. See the `{yx}px` fields of `ncinput`.

* 3.0.3 (2022-01-02)
  * No user-visible changes to the API, but Sixel quantization has been
    rewritten. It is now substantially faster, though quality has gone
    down for some images. I'll be working on bringing it back for 3.0.4.

* 3.0.2 (2021-12-21)
  * Added `ncplane_cursor_y()` and `ncplane_cursor_x()`.
  * Added `NCOPTION_SCROLLING`, equivalent to calling
    `ncplane_set_scrolling(true)` on the standard plane.
  * Added `NCOPTION_CLI_MODE`, an alias for the bitwise OR of
    `NCOPTION_SCROLLING`, `NCOPTION_NO_CLEAR_BITMAPS`,
    `NCOPTION_NO_ALTERNATE_SCREEN`, and `NCOPTION_PRESERVE_CURSOR`.
  * Added `ncvisual_from_sixel()`.
  * The control sequence corresponding to a pixel-blitted `ncvisual()`
    can be retrieved by using `ncplane_at_yx()` on the sprixel plane.

* 3.0.1 (2021-12-14)
  * Added the `NCPLANE_OPTION_VSCROLL` flag. Creating an `ncplane` with this
    flag is equivalent to immediately calling `ncplane_set_scrolling(true)`.
  * Added the `NCPLANE_OPTION_AUTOGROW` flag and the `ncplane_set_autogrow()`
    and `ncplane_autogrow_p()` functions. When autogrow is enabled, the plane
    is automatically enlarged to accommodate output at its right (no scrolling)
    or bottom (scrolling enabled) boundaries.
  * Added `notcurses_default_background()` and `notcurses_default_foreground()`.
  * Added `nccell_load_ucs32()`.
  * Added `nctree_add()` and `nctree_del()`. `nctree` is now dynamic.

* 3.0.0 (2021-12-01) **"In the A"**
  * Made the ABI/API changes that have been planned/collected during 2.x

rust/src/notcurses.rs
// functions already exported by bindgen : 36
// ------------------------------------------
//  notcurses_at_yx
//  notcurses_bottom
//# notcurses_canchangecolor
//# notcurses_canfade
//# notcurses_canopen_images
//# notcurses_canopen_videos
//# notcurses_cansixel
//# notcurses_cantruecolor
//# notcurses_canutf8
//  notcurses_cursor_disable
//  notcurses_cursor_enable
//# notcurses_debug
//  notcurses_drop_planes
//  notcurses_getc
//# notcurses_init
//  notcurses_inputready_fd
//  notcurses_lex_blitter
//  notcurses_lex_margins
//  notcurses_lex_scalemode
//  notcurses_mouse_disable
//  notcurses_mouse_enable
//  notcurses_palette_size
//  notcurses_refresh
//  notcurses_render
//  notcurses_render_to_buffer
//  notcurses_render_to_file
//  notcurses_stats
//  notcurses_stats_alloc
//  notcurses_stats_reset
//  notcurses_stdplane
//  notcurses_stdplane_const
//# notcurses_stop
//  notcurses_str_blitter
//  notcurses_str_scalemode
//  notcurses_supported_styles
//  notcurses_top
//  notcurses_ucs32_to_utf8
//  notcurses_version
//  notcurses_version_components
//
// static inline functions total: 6
// ----------------------------------------- (done / remaining)
// (+) implement : 5 / 1
// (#) unit tests: 0 / 6
// -----------------------------------------
//+ notcurses_align
//+ notcurses_getc_blocking
//+ notcurses_getc_nblock
//+ notcurses_stddim_yx
//+ notcurses_stddim_yx_const
//+ notcurses_term_dim_yx

use core::ptr::{null, null_mut};

use crate::{
    // NOTE: can't use libc::timespec nor libc::sigset_t
    // with notcurses_getc(()
    bindings::{sigemptyset, sigfillset, sigset_t, timespec},
    ncplane_dim_yx, notcurses_getc, notcurses_init, notcurses_stdplane, notcurses_stdplane_const,
    types::{
        NcAlign, NcInput, NcLogLevel, NcPlane, Notcurses, NotcursesOptions, NCALIGN_CENTER,
        NCALIGN_LEFT, NCOPTION_NO_ALTERNATE_SCREEN, NCOPTION_SUPPRESS_BANNERS,
    },
};

impl NotcursesOptions {
    /// Simple `NotcursesOptions` constructor
    pub fn new() -> Self {
        Self::with_all_options(0, 0, 0, 0, 0, 0)
    }

    /// `NotcursesOptions` constructor with customizable margins
    pub fn with_margins(top: i32, right: i32, bottom: i32, left: i32) -> Self {
        Self::with_all_options(0, top, right, bottom, left, 0)
    }

    /// `NotcursesOptions` constructor with customizable flags
    pub fn with_flags(flags: u64) -> Self {
        Self::with_all_options(0, 0, 0, 0, 0, flags)
    }

    /// `NotcursesOptions` constructor with all the options available
    ///
    /// ## Arguments
    ///
    /// - loglevel
    ///
    ///   Progressively higher log levels result in more logging to stderr. By
    ///   default, nothing is printed to stderr once fullscreen service begins.
    ///
    /// - margin_t, margin_r, margin_b, margin_l
    ///
    ///   Desirable margins (top, right, bottom, left).
    ///
    ///   If all are 0 (default), we will render to the entirety of the screen.
    ///   If the screen is too small, we do what we can.
    ///   Absolute coordinates are relative to the rendering area
    ///   ((0, 0) is always the origin of the rendering area).
    ///
    /// - flags
    ///
    ///   General flags; This is expressed as a bitfield so that future options
    ///   can be added without reshaping the struct.
    ///   Undefined bits must be set to 0.
    ///
    ///   - [`NCOPTION_INHIBIT_SETLOCALE`]
    ///   - [`NCOPTION_NO_ALTERNATE_SCREEN`]
    ///   - [`NCOPTION_NO_FONT_CHANGES`]
    ///   - [`NCOPTION_NO_QUIT_SIGHANDLERS`]
    ///   - [`NCOPTION_NO_WINCH_SIGHANDLER`]
    ///   - [`NCOPTION_SUPPRESS_BANNERS`]
    ///
    pub fn with_all_options(
        loglevel: NcLogLevel,
        margin_t: i32,
        margin_r: i32,
        margin_b: i32,
        margin_l: i32,
        flags: u64,
    ) -> Self {
        Self {
            termtype: null(),
            renderfp: null_mut(),
            loglevel,
            margin_t,
            margin_r,
            margin_b,
            margin_l,
            flags,
        }
    }
}

impl Notcurses {
    /// `Notcurses` simple constructor with clean output
    pub unsafe fn new<'a>() -> &'a mut Notcurses {
        let options = NotcursesOptions::with_flags(NCOPTION_SUPPRESS_BANNERS);
        &mut *notcurses_init(&options, null_mut())
    }

    /// `Notcurses` simple constructor, showing banners
    pub unsafe fn with_banners<'a>() -> &'a mut Notcurses {
        &mut *notcurses_init(&NotcursesOptions::new(), null_mut())
    }

    /// `Notcurses` simple constructor without an alternate screen
    pub unsafe fn without_altscreen<'a>() -> &'a mut Notcurses {
        let options = NotcursesOptions::with_flags(NCOPTION_NO_ALTERNATE_SCREEN);
        &mut *notcurses_init(&options, null_mut())
    }

    /// `Notcurses` simple constructor without an alternate screen
    pub unsafe fn without_altscreen_nor_banners<'a>() -> &'a mut Notcurses {
        let options =
            NotcursesOptions::with_flags(NCOPTION_NO_ALTERNATE_SCREEN | NCOPTION_SUPPRESS_BANNERS);
        &mut *notcurses_init(&options, null_mut())
    }

    /// `Notcurses` constructor with options
    pub unsafe fn with_options<'a>(options: &NotcursesOptions) -> &'a mut Notcurses {
        &mut *notcurses_init(options, null_mut())
    }
}

/// return the offset into 'availcols' at which 'cols' ought be output given the requirements of 'align'
#[inline]
pub fn notcurses_align(availcols: i32, align: NcAlign, cols: i32) -> i32 {
    if align == NCALIGN_LEFT {
        return 0;
    }
    if cols > availcols {
        return 0;
    }
    if align == NCALIGN_CENTER {
        return (availcols - cols) / 2;
    }
    availcols - cols // NCALIGN_RIGHT
}

/// 'input' may be NULL if the caller is uninterested in event details.
/// If no event is ready, returns 0.
// TODO: use pakr-signals
#[inline]
pub fn notcurses_getc_nblock(nc: &mut Notcurses, input: &mut NcInput) -> char {
    unsafe {
        let mut sigmask = sigset_t { __val: [0; 16] };
        sigfillset(&mut sigmask);
        let ts = timespec {
            tv_sec: 0,
            tv_nsec: 0,
        };
        core::char::from_u32_unchecked(notcurses_getc(nc, &ts, &mut sigmask, input))
    }
}

/// 'input' may be NULL if the caller is uninterested in event details.
/// Blocks until an event is processed or a signal is received.
#[inline]
pub fn notcurses_getc_nblocking(nc: &mut Notcurses, input: &mut NcInput) -> char {
    unsafe {
        let mut sigmask = sigset_t { __val: [0; 16] };
        sigemptyset(&mut sigmask);
        core::char::from_u32_unchecked(notcurses_getc(nc, null(), &mut sigmask, input))
    }
}

/// notcurses_stdplane(), plus free bonus dimensions written to non-NULL y/x!
#[inline]
pub fn notcurses_stddim_yx(nc: &mut Notcurses, y: &mut i32, x: &mut i32) -> NcPlane {
    unsafe {
        let s = notcurses_stdplane(nc);
        ncplane_dim_yx(s, y, x);
        *s
    }
}

/// notcurses_stdplane_const(), plus free bonus dimensions written to non-NULL y/x!
#[inline]
pub fn notcurses_stddim_yx_const(nc: &Notcurses, y: &mut i32, x: &mut i32) -> NcPlane {
    unsafe {
        let s = notcurses_stdplane_const(nc);
        ncplane_dim_yx(s, y, x);
        *s
    }
}

/// Return our current idea of the terminal dimensions in rows and cols.
#[inline]
pub fn notcurses_term_dim_yx(nc: &Notcurses, rows: &mut u32, cols: &mut u32) {
    unsafe {
        let mut irows = *rows as i32;
        let mut icols = *cols as i32;
        ncplane_dim_yx(
            notcurses_stdplane_const(nc), &mut irows, &mut icols);
    }
}

#[cfg(test)]
mod test {
    use serial_test::serial;

    use std::io::Read;

    use crate::{notcurses_stop, NcFile, Notcurses};

    /*
    #[test]
    #[serial]
    fn () {
    }
    */

    // Test the bindgen functions ----------------------------------------------

    #[test]
    #[serial]
    fn notcurses_init() {
        unsafe {
            let nc = Notcurses::new();
            assert![nc as *mut _ != core::ptr::null_mut()];
            notcurses_stop(nc);
        }
    }

    #[test]
    #[serial]
    #[ignore]
    // FIXME: always return null
    fn notcurses_at_yx() {
        unsafe {
            let nc = Notcurses::new();
            let mut sm = 0;
            let mut ch = 0;
            let res = crate::notcurses_at_yx(nc, 0, 0, &mut sm, &mut ch);
            notcurses_stop(nc);
            assert![!res.is_null()];

            //print!("[{}] ", res);
        }
    }

    #[test]
    #[serial]
    fn notcurses_debug() {
        unsafe {
            let nc = Notcurses::new();
            let mut _p: *mut i8 = &mut 0;
            let mut _size: *mut usize = &mut 0;
            let mut file = NcFile::from_libc(libc::open_memstream(&mut _p, _size));
            crate::notcurses_debug(nc, file.as_nc_ptr());
            notcurses_stop(nc);

            let mut string1 = String::new();
            let _result = file.read_to_string(&mut string1);

            let string2 =
                " ************************** notcurses debug state *****************************";

            assert_eq![&string1[0..string2.len()], &string2[..]];
        }
    }

    #[test]
    #[serial]
    fn notcurses_canchangecolor() {
        unsafe {
            let nc = Notcurses::new();
            let res = crate::notcurses_canchangecolor(nc);
            notcurses_stop(nc);
            print!("[{}] ", res);
        }
    }

    #[test]
    #[serial]
    fn notcurses_canfade() {
        unsafe {
            let nc = Notcurses::new();
            let res = crate::notcurses_canfade(nc);
            notcurses_stop(nc);
            print!("[{}] ", res);
        }
    }

    #[test]
    #[serial]
    fn notcurses_canopen_images() {
        unsafe {
            let nc = Notcurses::new();
            let res = crate::notcurses_canopen_images(nc);
            notcurses_stop(nc);
            print!("[{}] ", res);
        }
    }

    #[test]
    #[serial]
    fn notcurses_canopen_videos() {
        unsafe {
            let nc = Notcurses::new();
            let res = crate::notcurses_canopen_videos(nc);
            notcurses_stop(nc);
            print!("[{}] ", res);
        }
    }

    #[test]
    #[serial]
    fn notcurses_cansixel() {
        unsafe {
            let nc = Notcurses::new();
            let res = crate::notcurses_cansixel(nc);
            notcurses_stop(nc);
            print!("[{}] ", res);
        }
    }

    #[test]
    #[serial]
    fn notcurses_cantruecolor() {
        unsafe {
            let nc = Notcurses::new();
            let res = crate::notcurses_cantruecolor(nc);
            notcurses_stop(nc);
            print!("[{}] ", res);
        }
    }

    #[test]
    #[serial]
    fn notcurses_canutf8() {
        unsafe {
            let nc = Notcurses::new();
            let res = crate::notcurses_canutf8(nc);
            notcurses_stop(nc);
            print!("[{}] ", res);
        }
    }
}
